From c3ee3b4150feb2918d074a6b10b5a33328b091bf Mon Sep 17 00:00:00 2001
From: thampan <joby.thampan@intel.com>
Date: Fri, 27 Mar 2020 18:20:28 +0800
Subject: [PATCH] DRV_DPM_SW-210: BugFix for 1-Step PTP over UDP

    - BugFix For 1 Step PTP over UDP TSENALL register which was
      required to timestamp RX packets got overwritten
---
 drivers/net/ethernet/lantiq/switch-api/mac/xgmac_mac_api.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_mac_api.c b/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_mac_api.c
index 53d4caf3c194..037fe999cac8 100644
--- a/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_mac_api.c
+++ b/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_mac_api.c
@@ -1678,11 +1678,22 @@ int xgmac_ptp_txtstamp_mode(void *pdev,
 			    u32 tsevntena)
 {
 	u32 mac_tscr = 0;
+	struct mac_prv_data *pdata = GET_MAC_PDATA(pdev);
+
+	/* Read is required here, otherwise setting done by hwtstamp_ctrl
+	 * application will be overwritten
+	 */
+	mac_tscr = XGMAC_RGRD(pdata, MAC_TSTAMP_CR);
 
+	/* To Control Tx for 1-Step PTP over Eth or PTP over IPv4/IPv6 Tx packets
+	 * below 4 settings in MAC_TSTAMP_CR is required
+	 */
 	MAC_SET_VAL(mac_tscr, MAC_TSTAMP_CR, TSVER2ENA, 1);
 	MAC_SET_VAL(mac_tscr, MAC_TSTAMP_CR, TSIPENA, 1);
 	MAC_SET_VAL(mac_tscr, MAC_TSTAMP_CR, TSIPV4ENA, 1);
 	MAC_SET_VAL(mac_tscr, MAC_TSTAMP_CR, TSIPV6ENA, 1);
+
+	/* To say for 1-Step Only Insert timestamp for Sync Packets */
 	MAC_SET_VAL(mac_tscr, MAC_TSTAMP_CR, SNAPTYPSEL, snaptypesel);
 	MAC_SET_VAL(mac_tscr, MAC_TSTAMP_CR, TSEVNTENA, tsevntena);
 	MAC_SET_VAL(mac_tscr, MAC_TSTAMP_CR, TSMSTRENA, tsmstrena);
